### PR TITLE
store: correctly remove incomplete layers on load.

### LINKED ISSUE #2184: Can't run any podman command anymore - panic: runtime error: invalid memory address or nil pointer dereference
### Discussed in https://github.com/containers/podman/discussions/24736

<div type='discussions-op-text'>

<sup>Originally posted by **dozenposture** December  2, 2024</sup>
I've used to have about 20 containers running without issues.
While trying to update an image, I've found podman can't execute any command anymore.

It really doesn't matter which command I try to run, I always get the following error message:

```
podman@host:~> podman info
ERRO[0000] Refreshing container a2dea400161ecc6b8440e937e4fb745a6bda8fc87056ddc14c4289dfc560c6da: acquiring lock 2 for container a2dea400161ecc6b8440e937e4fb745a6bda8fc87056ddc14c4289dfc560c6da: file exists
WARN[0000] Found incomplete layer "6d1aab23cceb2aa47ce43e610416609023b6a0dc576c7b383b3eccf30ab6efd6", deleting it
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x100 pc=0x5633e1fec837]

goroutine 1 [running]:
github.com/containers/storage.(*layerStore).load(0xc000460280, 0x1)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/containers/storage/layers.go:917 +0xff7
github.com/containers/storage.(*store).newLayerStore(0x3b?, {0xc000049ef0, 0x28}, {0xc00060f8c0, 0x3b}, {0x0, 0x0}, {0x5633e327af18, 0xc000146d80}, 0x0)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/containers/storage/layers.go:1156 +0x6e5
github.com/containers/storage.(*store).getLayerStoreLocked(0xc0005ac000)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/containers/storage/store.go:1125 +0x25d
github.com/containers/storage.(*store).getLayerStore(0xc0005ac000)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/containers/storage/store.go:1141 +0x8f
github.com/containers/storage.(*store).writeToAllStores(0xc0005ac000, 0xc0005244c0)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/containers/storage/store.go:1401 +0x4c
github.com/containers/storage.(*store).DeleteContainer(0x93?, {0xc0005dcb00?, 0x0?})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/containers/storage/store.go:2703 +0x38
github.com/containers/podman/v5/libpod.(*storageService).DeleteContainer(0xc000116600, {0xc00060f500?, 0xc000260fc0?})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/libpod/storage.go:177 +0x72
github.com/containers/podman/v5/libpod.(*Container).teardownStorage(0xc0004b8780)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/libpod/container_internal.go:603 +0x314
github.com/containers/podman/v5/libpod.(*Runtime).removeContainer(0xc000505180, {0x5633e3267d90, 0x5633e4032600}, 0xc0004b8780, {0x0, 0x1, 0x0, 0x0, 0x0, 0x0, ...})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/libpod/runtime_ctr.go:981 +0x1888
github.com/containers/podman/v5/libpod.(*Runtime).refresh(0xc000505180, {0x5633e3267d90, 0x5633e4032600}, {0xc0005130c0, 0x1f})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/libpod/runtime.go:871 +0x52d
github.com/containers/podman/v5/libpod.makeRuntime({0x5633e3267d90, 0x5633e4032600}, 0xc000505180)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/libpod/runtime.go:637 +0x1bce
github.com/containers/podman/v5/libpod.newRuntimeFromConfig({0x5633e3267d90, 0x5633e4032600}, 0xc0000c4008, {0xc000525660, 0x0, 0xa?})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/libpod/runtime.go:213 +0x30e
github.com/containers/podman/v5/libpod.NewRuntime({0x5633e3267d90, 0x5633e4032600}, {0xc000525660, 0x0, 0x0})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/libpod/runtime.go:171 +0x5a
github.com/containers/podman/v5/pkg/domain/infra.getRuntime({0x5633e3267d90, 0x5633e4032600}, 0xc0004a4100, 0xc000525898)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/pkg/domain/infra/runtime_libpod.go:228 +0x1edc
github.com/containers/podman/v5/pkg/domain/infra.GetRuntime.func1()
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/pkg/domain/infra/runtime_libpod.go:46 +0x65
sync.(*Once).doSlow(0x1c?, 0xc0004be400?)
	/usr/lib64/go/1.23/src/sync/once.go:76 +0xb4
sync.(*Once).Do(...)
	/usr/lib64/go/1.23/src/sync/once.go:67
github.com/containers/podman/v5/pkg/domain/infra.GetRuntime({0x5633e3267d90?, 0x5633e4032600?}, 0x5633e2a79cfa?, 0xe?)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/pkg/domain/infra/runtime_libpod.go:45 +0x55
github.com/containers/podman/v5/pkg/domain/infra.NewLibpodImageRuntime(0xc0005259b0?, 0xc0004bbab8?)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/pkg/domain/infra/runtime_proxy.go:24 +0x27
github.com/containers/podman/v5/pkg/domain/infra.NewImageEngine(0x5633e2a79cfa?)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/pkg/domain/infra/runtime_abi.go:31 +0x4b
github.com/containers/podman/v5/cmd/podman/registry.NewImageEngine(0x5633e3f3b3c0?, {0x5633e4009e00?, 0x9?, 0x5633e2da6e58?})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/cmd/podman/registry/registry.go:50 +0x54
main.persistentPreRunE(0x5633e3f3b3c0, {0x5633e4032600, 0x0, 0x0})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/cmd/podman/root.go:317 +0x6f9
github.com/spf13/cobra.(*Command).execute(0x5633e3f3b3c0, {0xc0000400d0, 0x0, 0x0})
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/spf13/cobra/command.go:956 +0x931
github.com/spf13/cobra.(*Command).ExecuteC(0x5633e3f23e00)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/spf13/cobra/command.go:1117 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/spf13/cobra/command.go:1041
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/vendor/github.com/spf13/cobra/command.go:1034
main.Execute()
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/cmd/podman/root.go:116 +0xb4
main.main()
	/home/abuild/rpmbuild/BUILD/podman-5.3.1/cmd/podman/main.go:61 +0x4b2
```

I've browsed for similar issues and I've found something about lingering. The podman user has lingering enabled (since the beginning)

Can someone please help me out understanding what went wrong and how to fix this?
Thanks!</div>

### PR DESCRIPTION
**Pull request by Luap99**
In go one should never modify a slice while also iterating over it at the same time. This causes weird side effects as the underlying array elements are shifted around without the range loop index knowing. So if you delete a element the loop will then actually skip the next one and theoretically access out of bounds on the last element which does not panic but rather return the default zero type, nil here which then causes the panic on layer.Flags == nil.

Here is a simple example to show the behavior:
```
func main() {
	slice := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	for _, num := range slice {
		if num == 5 {
			slice = slices.DeleteFunc(slice, func(n int) bool {
				return n == 5
			})
		}
		fmt.Println(num)
	}
}
```

The loop will not print 6, but then as last number it prints 0 (the default zero type for an int).

Fixes #2184

### CODE REVIEW

**Comment by Luap99** [comment]
@giuseppe @mtrmac PTAL

Untested at this point as I am not familiar with the storage layout so I am not sure how I can corrupt my storage in such case to trigger this code path.

I have also not yet audited other code paths for a similar issues. I wonder if there is a linter for this thing.

File: layers.go [hunk 1]
```diff
@@ -918,18 +919,26 @@ func (r *layerStore) load(lockedForWriting bool) (bool, error) {
...
-					// deleted incomplete layers have their metadata correctly removed.
-					incompleteDeletionErrors = multierror.Append(incompleteDeletionErrors,
-						fmt.Errorf("deleting layer %#v: %w", layer.ID, err))
-				}
+				// Important: Do not call r.deleteInternal() here. It modifies r.layers
+				// which causes unexpected side effects while iterating over r.layers here.
+				// The range loop has no idea that the underlying elements where shifted
+				// around.
+				layersToDeleteIDs = append(layersToDeleteIDs, layer.ID)
 				modifiedLocations |= layerLocation(layer)
```
Comment (cgwalters) [inline layers.go:927]: Even though it's ready only maybe this function call move too just for good measure?

Then conceptually of course this loop boils down to what in Rust would just be `r.layers.iter().filter(|layer| layer.is_incomplete())`

**Review by cgwalters — APPROVED** [review]
Can't resist noting this type of bug can't happen in Rust.

File: layers.go [hunk 2]
```diff
@@ -905,6 +905,7 @@ func (r *layerStore) load(lockedForWriting bool) (bool, error) {
 		// and possibly loadMounts() again.
 	}
 
+	var layersToDeleteIDs []string
```
Comment (mtrmac) [inline layers.go:908]: Move this immediately next to `incompleteDeletionErrors`, please — both variables are local to the `errorToResolveBySaving` case, and processed ~together.

File: layers.go (hunk 1 above)
Comment (mtrmac) [reply layers.go:927]: I agree that moving this to the new loop would be cleaner. We would then need to collect `*Layer` pointers instead, but that should be fine; the loading code above has already ensured layer IDs are unique.

**Review by mtrmac — COMMENTED** [review]
Thanks, great find!

**Comment by mtrmac** [comment]
> Untested at this point as I am not familiar with the storage layout so I am not sure how I can corrupt my storage in such case to trigger this code path.

In this case, probably by deleting an image (or a single layer): that marks layers as incomplete and proceeds to delete the layer files. If the deletion is aborted (by panic / reboot / SIGKILL) at that time, the layer remains as incomplete and triggers this late deletion. (It would also be necessary to arrange that the layer being deleted is not the last one in `r.layers`, or the loop would terminate.)

I don’t think there is a way to write a _reliable_ test for this (using the stable API); the caller would have to trigger deletion of a layer in one thread, and abort the process in another, _hoping_ that it aborts it during the filesystem deletion. (Alternatively, a hack would be to call `SetFlag` with the incomplete flag explicitly; that should not be too hard to trigger, but it’s not a public API, and it seems to me like a rather specific scenario to write a test for.)

File: layers.go [hunk 3]
```diff
@@ -913,20 +913,29 @@ func (r *layerStore) load(lockedForWriting bool) (bool, error) {
...
+		for _, layer := range layersToDelete {
+			logrus.Warnf("Found incomplete layer %q, deleting it", layer.ID)
+			err := r.deleteInternal(layer.ID)
+			if err != nil {
+				// Don't return the error immediately, because deleteInternal does not saveLayers();
+				// Even if deleting one incomplete layer fails, call saveLayers() so that other possible successfully
+				// deleted incomplete layers have their metadata correctly removed.
+				incompleteDeletionErrors = multierror.Append(incompleteDeletionErrors,
+					fmt.Errorf("deleting layer %#v: %w", layer.ID, err))
 				modifiedLocations |= layerLocation(layer)
```
Comment (mtrmac) [inline layers.go:939]: This needs to happen outside the `err != nil` condition.

LGTM otherwise.

Comment (Luap99) [reply layers.go]: ah yes, bad copy paste. fixed

**Review by mtrmac — COMMENTED** [review]
/lgtm

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by EvVlF** [comment]
[link back on #24843](https://github.com/containers/podman/issues/24843)
I delete all from `$HOME/.local/share/containers/storage/` and `$HOME/.local/share/containers/cache/` (sobbing with pain in one spot) - aaaand.. it didn't work(
Then, I built the program from the source code. Now, i have a version 5.4.0-dev.
Buuuuut, 
```

podman.service: Failed with result 'exit-code'.
systemd

podman.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
systemd

/home/koshon/builds/podman/podman/cmd/podman/main.go:61 +0x4b2
podman

main.main()
podman

/home/koshon/builds/podman/podman/cmd/podman/root.go:116 +0xb4
podman

main.Execute()
podman

/home/koshon/builds/podman/podman/vendor/[github.com/spf13/cobra/command.go:1034](http://github.com/spf13/cobra/command.go:1034)
podman

[github.com/spf13/cobra.(*Command).ExecuteContext(...)](http://github.com/spf13/cobra.(*Command).ExecuteContext(...))
podman

/home/koshon/builds/podman/podman/vendor/[github.com/spf13/cobra/command.go:1041](http://github.com/spf13/cobra/command.go:1041)
podman

[github.com/spf13/cobra.(*Command).Execute(...](http://github.com/spf13/cobra.(*Command).Execute(...))
podman
```
(sorry, this log from cockpit web console)
but if you need some other logs - tell me which ones you want to output, by what command.

```
go version go1.23.4 linux/amd64
```
When I built from source, there were no errors (I mean with 'make')
[Performed according to this manual](https://podman.io/docs/installation#building-from-source)
what are my next steps - I can't wait until February((

**Pull request merged**